src/transaction/log_recovery.c

```diff
@@ -6389,85 +6389,32 @@ log_rv_pack_undo_record_changes (char *ptr, int offset_to_data, int old_data_siz
  * return        : fixed page or NULL
  * thread_p (in) : thread entry
  * vpid_rcv (in) : page identifier
- * rcvindex (in) : recovery index of log record to redo
  */
 STATIC_INLINE PAGE_PTR
-log_rv_redo_fix_page (THREAD_ENTRY * thread_p, const VPID * vpid_rcv, LOG_RCVINDEX rcvindex)
+log_rv_redo_fix_page (THREAD_ENTRY * thread_p, const VPID * vpid_rcv)
 {
   PAGE_PTR page = NULL;
 
   assert (vpid_rcv != NULL && !VPID_ISNULL (vpid_rcv));
 
-  /* how it works:
-   * since we are during recovery, we don't know the current state of page. it may be unreserved (its file is destroyed)
-   * or it may not be allocated. these are expected cases and we don't want to raise errors if it happens.
-   * some redo records are used to initialize a page for the first time (also setting its page type which is necessary
-   * to consider a page allocated). even first attempt to fix page fails, but the page's sector is reserved, we will
-   * fix the page as NEW_PAGE and apply its initialization redo log record.
-   * In case of RVPGBUF_COMPENSATE_DEALLOC, we expect deallocated page.
-   */
-
-  if (rcvindex == RVPGBUF_COMPENSATE_DEALLOC)
-    {
-      page = pgbuf_fix (thread_p, vpid_rcv, OLD_PAGE_DEALLOCATED, PGBUF_LATCH_WRITE, PGBUF_UNCONDITIONAL_LATCH);
-      if (page == NULL)
-	{
-	  assert_release (false);
-	  return NULL;
-	}
-
-      return page;
-    }
-
-  /* let's first try to fix page if it is not deallocated. */
-  if (pgbuf_fix_if_not_deallocated (thread_p, vpid_rcv, PGBUF_LATCH_WRITE, PGBUF_UNCONDITIONAL_LATCH, &page)
-      != NO_ERROR)
+  //
+  // during recovery, we don't care if a page is deallocated or not, apply the changes regardless. since changes to
+  // sector reservation table are applied in parallel with the changes in pages, at times the page may appear to be
+  // deallocated (part of an unreserved sector). but the changes were done while the sector was reserved and must be
+  // re-applied to get a correct end result.
+  // 
+  // moreover, the sector reservation check is very expensive. running this check on every page fix costs much more
+  // than any time gained by skipping redoing changes on deallocated pages.
+  //
+  // therefore, fix page using RECOVERY_PAGE mode. pgbuf_fix will know to accept even new or deallocated pages.
+  //
+  page = pgbuf_fix (thread_p, vpid_rcv, RECOVERY_PAGE, PGBUF_LATCH_WRITE, PGBUF_UNCONDITIONAL_LATCH);
```

Comment:
In `pgbuf_claim_bcb_for_fix()` called by `pgbuf_fix()`, if the `fetch_mode` is not `NEW_PAGE`, it tries to read the page from a disk volume. From what I've understood, It means that it would read a data page from a volume even though the recovery index is `RCV_IS_NEW_PAGE_INIT` here. It seems that it is unacceptable. Could I have your opinion about it?

Reply:
It should read the page from the volume (disk) regardless of recovery index.
Upon reading the page from disk, there are subsequent checks (ie: in `log_rv_fix_page_and_check_redo_is_needed`) to decide whether it needs to apply the log record or not, based on the LSA of the page and the LSA of the log record.

Log record with recovery indexes in `RCV_IS_NEW_PAGE_INIT` have no guarantee that the state of the page on disk is de-allocated. But the LSA check covers these log records too:
- If the page is never allocated, it has null LSA;
- If the page is de-allocated, its LSA must be smaller than the log record's; as a result, the record is redone
- If the page is not de-allocated, then its LSA must be equal or higher and the log record must not be redone

Note that previous code also read the page from disk by calling `pgbuf_fix_if_not_deallocated` first. If the page was not allocated, the pgptr remained NULL and pgbuf_fix was called again with `NEW_PAGE` flag. Although the approach was different, the end result was the same.

Reply:
Oh, I missed it reads a data page even with recovery indexes in `RCV_IS_NEW_PAGE_INIT` through `pgbuf_fix_if_not_deallocated()`. So, no additional I/O read. Thank you.

I think the code is correct, but have a little different thought about it. So, to understand correctly, I'd love it if you point out what I misunderstand.

> * If the page is de-allocated, its LSA must be smaller than the log record's; as a result, the record is redone

It seems that the LSA of a page could be larger than log recods even if the page is de-allocated. Imagine operations: (1) allocate - (2) deallocate - (3) allocate - (4) deallocate. If the redo recovery meets the log record of (1), but the LSA of the page is (4), it can be larger.

> * If the page is not de-allocated, then its LSA must be equal or higher and the log record must not be redone

As same above, the LSA could be smaller.

The way I see it, we can skip reading a data page and apply log records if its recovery index is in `RCV_IS_NEW_PAGE_INIT`. Even if the LSA of a data page was larger than the LSA of the log record, the subsequent redo operations will make it catch up on the latest state again. However, we can skip this redundant redo with your way. 

Please let me have your opinion.

Reply:
I think you are right that the examples given above in the second and third bullets are incomplete (or, maybe, plain wrong).
I do agree with your remark that a page can find itself in any position before, in-between or after with respect to the states (1)-(2)-(3)-(4) described above.
However, the verification:
```
		  if (LSA_LE (&rcv_lsa, rcv_page_lsaptr))
		    {
		      /* It is already done */
		      pgbuf_unfix (thread_p, rcv.pgptr);
		      break;
		    }
```
which is done for after all calls to `log_rv_redo_fix_page` (and which, on the branch `scalability_dev` from where we back-ported this change is abstracted away in an explicit function call - `log_rv_fix_page_and_check_redo_is_needed` at https://github.com/CUBRID/cubrid/blob/85101805863dfc12471e97632f2f944177c4830d/src/transaction/log_recovery.c#L430) covers all the cases that you describe.
Another check to avoid applying the log record would only be justified if it would improve performance even further, which, at this point is unknown.

Reply:
I agree that we don't need the additional check necessarily. I just wanted to check if I understand correctly. Thank you.